Add SmbiosConfigDxe to perform Runtime Smbios Table Overrides

christopherco opened the pull request:
One of the manufacturing requests was to support a method to dynamically
change specific Smbios Table data from a file present in the EFI partition.

This patch adds a new Dxe driver that, on first boot, will find a JSON file called
Smbios.cfg in the EFI partition. The driver will parse the JSON and add Key/Value
pairs representing the Smbios String field to override. The driver will then
save these as NV variables.

This patch also enlightens the Smbios driver to check for the presence of these
NV override variables and places them in the Smbios table.

christopherco committed:
Initial commit for SmbiosConfigDxe

christopherco committed:
Enable file read from SD card

This patch set DXE dependencies to load PlatformSmbiosDxe after
SmbiosConfig. It also adds support to read a specific file from
the SD Card's EFI partition.

christopherco committed:
Add json parser

This commit adds code to parse a simplified json file. The json file must
be only one level and all key/value pairs must be encapsulated by quotation marks.
We use this json file to populate key/value nodes in a linked list which will
be written to NV store.

christopherco committed:
Add Variable support

christopherco committed:
Read Smbios Overrides from NV

christopherco commented on Silicon/NXP/iMX6Pkg/Drivers/PlatformSmbiosDxe/PlatformSmbiosDxe.c:
```diff
@@ -46,11 +46,94 @@
 #include "PlatformSmbiosDxe.h"
 
+// TODO: Move to common header
+#define MAX_VARIABLE_SIZE 100
```
Move to common library?

christopherco commented on Silicon/NXP/iMX6Pkg/Drivers/PlatformSmbiosDxe/PlatformSmbiosDxe.c:
```diff
@@ -46,11 +46,94 @@
+
+  Retrieve an SMBIOS Override variable from UEFI variables.
+
+  @param[in] VariableName         Pointer to the string containing the key name.
```
Fix these comments

christopherco commented on Silicon/NXP/iMX6Pkg/Drivers/PlatformSmbiosDxe/PlatformSmbiosDxe.c:
```diff
@@ -46,11 +46,94 @@
+  return Status;
+}
+
+EFI_STATUS
```
Add comment header

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,712 @@
+
+#include "SmbiosConfigDxe.h"
+
+typedef enum JsonParserState {
```
Move to private header

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,712 @@
+    goto Exit;
+  }
+
+  for (i = 0; i < MAX_SMBIOS_KEY_REV_1; i++) {
```
Don't need to do this check. Just insert the values that we care about into the list.

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,712 @@
+    goto Exit;
+  }
+
+  Status = File->Read (
```
Limit open/read size to prevent huge file sizes?

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,712 @@
+  Status = gRT->SetVariable (Key,
+                &giMXPlatformSmbiosOverrideGuid,
+                EFI_VARIABLE_NON_VOLATILE | EFI_VARIABLE_BOOTSERVICE_ACCESS |
+                EFI_VARIABLE_RUNTIME_ACCESS,
```
Remove RUNTIME_ACCESS

christopherco commented on Silicon/NXP/iMX6Pkg/iMX6CommonDsc.inc:
```diff
@@ -924,6 +924,11 @@
+  #
+  # Dynamic Smbios Override
+  #
+  Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.inf
```
Add CONFIG option to disable this override driver

christopherco commented on Silicon/NXP/iMX6Pkg/iMX6CommonFdf.inc:
```diff
@@ -48,6 +48,11 @@
+  #
+  # Dynamic Smbios Override
+  #
+  INF Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.inf
```
Add CONFIG option to disable this override driver

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,712 @@
+  TempKey = NULL;
+  TempValue = NULL;
+
+  while (CurrentOffset < BufferSize) {
```
Refactor Json Parser into its own library?

christopherco replied:
Also this simplified json is probably a bit excessive. Could achieve the same result using a simple CSV parser. Consider changing over once the parsing code is in its own library.

christopherco commented on Silicon/NXP/iMX6Pkg/Drivers/PlatformSmbiosDxe/PlatformSmbiosDxe.inf:
```diff
@@ -81,4 +83,5 @@
 
 [Depex]
-  gEfiSmbiosProtocolGuid
+  AFTER gSmbiosConfigDxeFileGuid
```
Add the CONFIG here too.

christopherco committed:
Add PCD to enable override

christopherco committed:
Remove runtime access to SmbiosConfig variables

christopherco committed:
Update comments

christopherco committed:
Minor refactoring

christopherco committed:
Initialize overrides to NULL to prevent bad FreePool

christopherco committed:
Add common header

christopherco commented on Silicon/NXP/iMXPlatformPkg/Include/iMXVarStore.h:
```diff
@@ -0,0 +1,20 @@
+
+#define MAX_VARIABLE_SIZE 100
+
+#endif // _IMX_VAR_STORE_H_
```
nit - Need new line

dmcilvaney commented on Silicon/NXP/iMXPlatformPkg/Include/iMXVarStore.h:
```diff
@@ -0,0 +1,20 @@
+#ifndef __IMX_VAR_STORE_H__
+#define __IMX_VAR_STORE_H__
+
+#define MAX_VARIABLE_SIZE 100
```
SMBIOS 6.1.3 allows strings to be unlimited length, probably not a big issue but we can easily allow at least up to 16k-(size of meta data) which is the OP-TEE max variable size.

neilsh-msft commented on Silicon/NXP/iMX6Pkg/Drivers/PlatformSmbiosDxe/PlatformSmbiosDxe.c:
```diff
@@ -617,50 +694,78 @@ SysInfoUpdateSmbiosType1 (
-    DEBUG ((DEBUG_ERROR, "%a: PcdSystemManufacturer not filled\n", __FUNCTION__));
-    Status = EFI_INVALID_PARAMETER;
-    goto Exit;
+  Status = RetrieveSmbiosVariable (L"System Manufacturer",
```
Should the "System Manufacturer" and other const string identifiers themselves be defined as a pcd value, or defined in a header file?

christopherco replied:
Yes. Common header makes sense here. Will fix

neilsh-msft commented on Silicon/NXP/iMX6Pkg/Drivers/PlatformSmbiosDxe/PlatformSmbiosDxe.c:
```diff
@@ -617,50 +694,78 @@ SysInfoUpdateSmbiosType1 (
+                                   &ManufacturerOverrideLen);
+  if (Status == EFI_SUCCESS) {
+    Manufacturer = ManufacturerOverride;
+    ManufacturerLen = ManufacturerOverrideLen;
```
Can this be zero length?

christopherco replied:
Good catch. It cannot not be 0 length otherwise we get \0\0 which signifies the end of the string area. Will change

neilsh-msft commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,705 @@
+
+  If we successfully find both Key and Value positions and lengths, then allocate/copy
+  these to their respective buffers and return to the caller.
+
```
@param declarations?

neilsh-msft commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,705 @@
+  CHAR8 *TempKey;
+  CHAR8 *TempValue;
+
+  CurrentOffset = *Offset;
```
Check Offset, Buffer, and other parameters for NULL and return bad parameter error?

neilsh-msft commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,705 @@
+  *Offset = CurrentOffset;
+
+Exit:
+  return Status;
```
Should free memory resources on error.

neilsh-msft commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,705 @@
+        if (ValueStart == NULL) {
+          ValueStart = &Buffer[CurrentOffset];
+        }
+        if (Buffer[CurrentOffset] == '"') {
```
Values can't contain quote characters.  No escaping mechanism.

christopherco replied:
Discussed this with Daniel a bit. Seems like the best move is to simplify to an INI parser with just simple key=value\n setup and remove this JSON parsing attempt. The INI parser will treat the Value as a literal so no escape characters needed.
Sample:
```
Signature=SmbiosOverrideConfigurationTable
Revision=1
System Manufacturer=Contoso
System Family=Contoso Family
```

neilsh-msft commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,705 @@
+  // Fill Linked List with File data
+  Offset = 0;
+  Count = 0;
+  Status = GetNextJsonKeyword(Buffer, &Offset, BufferSize, &NextChar);
```
Check parameter values and return invalid parameter on error.

hbeberman commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.h:
```diff
@@ -0,0 +1,47 @@
+  StateKeyDone,
+  StateFindValue,
+  StateReadValue,
+  StateValueDone,  
```
nit: Extra whitespace

hbeberman commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,705 @@
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+  
```
nit: There are several lines with extra whitespace throughout the file.

hbeberman commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,705 @@
+    Status = EFI_OUT_OF_RESOURCES;
+    goto Exit;
+  }
+  CopyMem(TempKey, KeyStart, KeyLength);
```
Add a space after function name. There are several more instances throughout the file.

christopherco replied:
Please point them out. I already did a scrub and missed them

hbeberman commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,705 @@
+    DEBUG ((DEBUG_INFO, "%a: Found %s\n", __FUNCTION__, UnicodeDebugString));
+  }
+  Status = EFI_SUCCESS;
+Exit:
```
Empty line before the Exit flag for consistency.

christopherco committed:
Address review comments

christopherco committed:
Update linked list to store unicode. Add comments

christopherco commented on Silicon/NXP/iMX6Pkg/Drivers/PlatformSmbiosDxe/PlatformSmbiosDxe.c:
```diff
@@ -32,24 +32,101 @@
+**/
+EFI_STATUS
+RetrieveSmbiosVariableByGuid (
+  CONST CHAR16 *VariableName,
```
IN/OUT

christopherco commented on Silicon/NXP/iMX6Pkg/Drivers/PlatformSmbiosDxe/PlatformSmbiosDxe.c:
```diff
@@ -32,24 +32,101 @@
+**/
+EFI_STATUS
+RetrieveSmbiosVariable (
+  CONST CHAR16 *VariableName,
```
IN/OUT

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,879 @@
+  IN CHAR8 *Buffer,
+  IN OUT UINTN *Offset,
+  IN UINTN BufferSize,
+  CHAR16 **Key,
```
OUT

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,879 @@
+EFI_STATUS
+EFIAPI
+GetSmbiosOverride (
+  CONST CHAR16 *Key,
```
IN/OUT

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,879 @@
+EFI_STATUS
+EFIAPI
+PopulateSmbiosOverrideList (
+  CHAR8* Buffer,
```
IN/OUT

christopherco commented on Silicon/NXP/iMXPlatformPkg/Include/iMXVarStore.h:
```diff
@@ -0,0 +1,31 @@
+#define IMX_VARIABLE_SMBIOS_FAMILY              L"System Family"
+#define IMX_VARIABLE_SMBIOS_BASEBOARD_PRODUCT   L"Baseboard Product"
+
+#endif // _IMX_VAR_STORE_H_
```
__IMX_VAR_STORE_H__

christopherco committed:
Update style

christopherco committed:
Add INI validation code

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,878 @@
+  @param[out]     Key           Pointer to Unicode string containing Key
+  @param[out]     Value         Pointer to Unicode string containing Value
+
+  @retval EFI_SUCCESS             
```
Fill in retval comments

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,878 @@
+
+  // Bounds check on Key and Value
+  if ((KeyLength == 0) || (ValueLength == 0)) {
+    Status = EFI_INVALID_PARAMETER;
```
Should be EFI_NOT_FOUND

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,878 @@
+  }
+
+  if ((KeyLength > MAX_VARIABLE_SIZE) || (ValueLength > MAX_VARIABLE_SIZE)) {
+    Status = EFI_INVALID_PARAMETER;
```
Does not need to be an error. Just truncate.

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,878 @@
+  // Loop until end of file or we find a valid key/value pair
+  //
+  while (CurrentOffset < BufferSize) {
+    if (State == StateValueDone) {
```
Corner case: we don't return the final key/value pair of a valid INI file because we don't ever enter this check after the last \n is seen.

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,878 @@
+  //
+  // Scan through the INI file and add its contents into the SmbiosOverride linked list
+  //
+  Status = PopulateSmbiosOverrideList (Buffer, BufferSize);
```
Add a function to validate the INI syntax itself before populating the override list.
Otherwise you end up with a partially filled override list.

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,878 @@
+    goto Exit;
+  }
+
+  BufferSize = (UINTN) FileInfo->FileSize + sizeof (CHAR16);
```
Bounds check the FileSize. Expected file sizes are <1K Bytes. So limit File size to 10x this.

christopherco commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,878 @@
+  return Status;
+}
+
+
```
Remove this newline

christopherco committed:
Fix corner case

Corner case: we don't return the final key/value
pair since the final \n occurs at the end of the
file

christopherco committed:
Add max file size limit

christopherco committed:
More cleanup

dmcilvaney commented on Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c:
```diff
@@ -0,0 +1,945 @@
+*  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+*
+**/
+
```
Might be nice to have a "why does this exist" section here for future reference, briefly explaining the general flow of how SMBIOS override works (including Henry's provisioning flow as well maybe?) and why we did it. Or maybe in the PlatformSMBIOS as an alternative location?

christopherco replied:
Added

christopherco committed:
Increase max variable size to 1KB

christopherco committed:
Add driver comment
